Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3678,6 +3682,30 @@ std::pair<B3::PatchpointValue*, PatchpointExceptionHandle> AirIRGenerator::emitC
     return { patchpoint, exceptionHandle };
 }
 
+std::pair<B3::PatchpointValue*, PatchpointExceptionHandle> AirIRGenerator::createReturnCallPatchpoint(BasicBlock* block, Checked<int32_t>& newStackOffset, Vector<ArgumentLocation>& locations, const Vector<TypedTmp>& args, Vector<ConstrainedTmp> extraArgs)
+{
+    auto* patchpoint = addPatchpoint(B3::Void);
+    patchpoint->effects.terminal = true;
+    patchpoint->clobberEarly(RegisterSet::macroScratchRegisters());
+    patchpoint->clobberLate(RegisterSet::volatileRegistersForJSCall());
+    RegisterSet calleeSaveRegisters = RegisterSet::calleeSaveRegisters();
+    calleeSaveRegisters.exclude(RegisterSet::stackRegisters());
+    patchpoint->clobber(calleeSaveRegisters);
+
+    // NOTE: We reuse the caller stack space and spill the exceeding arguments to avoid corrupting slots below the frame pointer. They will be moved to the correct locations in the generator.
+    for (unsigned i = 0; i < args.size(); ++i) {
+        Tmp tmp = args[i];
+        if (locations[i].isStackArgument() && newStackOffset + locations[i].offsetFromSP() > CallFrame::returnPCOffset()) {
```

Comment:
How does this work? returnPCOffset is measured from the frame pointer, not from SP, right?

Reply:
But these locations are still relative to the caller's SP, right? so Callee FP = Caller FP, and newStackOffset tracks the difference between the SPs? Hmm, if we measure everything from FP, would that be simpler?

Reply:
I think the reasoning **is** from FP here.
`newStackOffset` is an offset from FP that gives us the future stack address, after shuffling.
Still from the caller's FP, `newStackOffset + returnPCOffset` represents the offset of the preserved return address, after shuffling.
Locations are indeed from the caller's SP. 
`locations.offsetFromSP + newStackOffset` gives us the offset of the stack arguments after the shuffling, but from the current FP since we are still in the caller.
The check is here to prevent clobbering anything below FP+returnPCOffset (i.e. callee saves and stack slots that might be used by the register allocator).
If we can reuse the stack slots that currently holds the caller's arguments (above FP), we do it in place, otherwise we pass them as arguments to the patchpoint, and the register allocator will put them at the top of the stack. We will move them to their new offsets in the patchpoint generator, since at this point we have restored the callee saves and the register allocator is done. This only happens when the callee requires more stack space than the caller did.